### PR TITLE
get rid of If in favor of Compare+Lookup

### PR DESCRIPTION
**Pull request by avibryant**
The new `Compare` node more directly models the JVM instruction set, and when combined with Lookup gives us what should be a more efficient implementation of branching on comparisons than we had previously, and with fewer primitives. Still needs a bit of cleanup and validation via benchmarks that it is in fact faster.

### CODE REVIEW

File: rainier-core/src/main/scala/com/stripe/rainier/core/Categorical.scala [hunk 1]
```diff
@@ -99,9 +99,7 @@ object Multinomial {
     Combinatorics.factorial(multi.k) + Real.sum(v.toList.map {
       case (t, i) =>
         val p = multi.pmf.getOrElse(t, Real.zero)
-        val pTerm =
-          If(i, i * p.log, If(p, whenNonZero = i * p.log, whenZero = Real.zero))
-
+        val pTerm = i * p.log //???
```
Comment (mio-stripe) [inline rainier-core/src/main/scala/com/stripe/rainier/core/Categorical.scala]: Looking at this again, it took me a while to refigure out that `pmf` is the map of outcomes to their probabilities and `v` is the map of outcomes to their number of occurrences. An extra sentence in the comment on the role of `value` in the `logDensity` would help.

File: rainier-benchmark/benchmarks.txt [hunk 2]
```diff
@@ -122,11 +121,11 @@ run                  Geometric                 100                            1.
 run                  Geometric                 1000                           1.000
 run                  Geometric                 10000                          1.000
 run                  Geometric                 100000                         1.000
-run                  GeometricZeroInflated     100                            1.000
```
Comment (avibryant) [inline rainier-benchmark/benchmarks.txt:125]: why'd this get slower?

**Comment by avibryant** [comment]
I think this is worth merging at this point, though could certainly use more documentation and optimization.

**Comment by mio-stripe** [comment]
Whoops, forgot to check back on this. Looking now.

**Comment by mio-stripe** [comment]
Cool, LGTM

**Comment by mio-stripe** [comment]
Looks like just the expected SBCs to update

**Pull request merged**